### PR TITLE
Set config-controller version label on CM App CRs

### PR DESCRIPTION
**Pull request by kopiczko**
Towards https://github.com/giantswarm/giantswarm/issues/15075

### CODE REVIEW

**Comment by kopiczko** [comment]
Back to draft. I need to rethink couple of things and add tests.

**Comment by kopiczko** [comment]
Now it's ready.

File: pkg/app/mutate_app.go [hunk 1]
```diff
@@ -76,10 +76,12 @@ func (m *Mutator) Mutate(request *v1beta1.AdmissionRequest) ([]mutator.PatchOper
...
 	m.logger.Debugf(ctx, "mutating app %#q in namespace %#q", appNewCR.Name, appNewCR.Namespace)
 
-	// We check the deletion timestamp because app CRs may be deleted by
-	// deleting the namespace they belong to.
 	if request.Operation == v1beta1.Update && !appNewCR.DeletionTimestamp.IsZero() {
-		m.logger.Debugf(ctx, "admitted deletion of app %#q in namespace %#q", appNewCR.Name, appNewCR.Namespace)
+		m.logger.Debugf(ctx, "skipping mutation for UPDATE operation of app %#q in namespace %#q with non-zero deletion timestamp", appNewCR.Name, appNewCR.Namespace)
+		return nil, nil
+	}
+	if request.Operation == v1beta1.Connect {
```
Comment (kopiczko) [inline pkg/app/mutate_app.go:83]: Ignoring connect.

**Review by kubasobon — APPROVED** [review]

File: pkg/app/mutate_app.go (hunk 1 above)
Comment (rossf7) [reply pkg/app/mutate_app.go:83]: Wow nice catch. I didn't know connect operations were a thing.

The new message is also clearer about the deletion timestamp. 👍

File: pkg/app/mutate_app.go [hunk 2]
```diff
@@ -94,13 +96,21 @@ func (m *Mutator) Mutate(request *v1beta1.AdmissionRequest) ([]mutator.PatchOper
...
+	// Set empty labels and annotations in case they are not set. This is
+	// in case we add new entires to null JSON objects. We don't want to do
+	// this as needed because it can be potentially overwritten if set
+	// after other patches.
+	if len(app.Annotations) == 0 {
+		result = append(result, mutator.PatchAdd("/metadata/annotations", map[string]string{}))
+	}
+	if len(app.Labels) == 0 {
+		result = append(result, mutator.PatchAdd("/metadata/labels", map[string]string{}))
+	}
```
Comment (rossf7) [inline pkg/app/mutate_app.go:112]: Nice to handle this at the start. 👍

File: pkg/app/mutate_app_test.go [hunk 3]
```diff
@@ -236,11 +248,150 @@ func Test_MutateApp(t *testing.T) {
 			configMaps: []*corev1.ConfigMap{
 				newTestConfigMap("other-app-values", "eggs2"),
 			},
+			operation: v1beta1.Create,
+			expectedPatches: []mutator.PatchOperation{
+				mutator.PatchAdd("/metadata/annotations", map[string]string{}),
+			},
+		},
+		{
+			name: "case 6: set app-operator paused annotation and config-controller version label for CM app on CREATE",
```
Comment (rossf7) [inline pkg/app/mutate_app_test.go:257]:
```suggestion
			name: "case 6: set app-operator paused annotation and config-controller version label for MC app on CREATE",
```

Shouldn't this be MC?

File: pkg/app/mutate_app.go [hunk 4]
```diff
@@ -94,13 +96,21 @@ func (m *Mutator) Mutate(request *v1beta1.AdmissionRequest) ([]mutator.PatchOper
...
 func (m *Mutator) MutateApp(ctx context.Context, app v1alpha1.App, operation v1beta1.Operation) ([]mutator.PatchOperation, error) {
-	if operation == v1beta1.Connect {
-		return nil, nil
-	}
 
 	var err error
 	var result []mutator.PatchOperation
 
+	// Set empty labels and annotations in case they are not set. This is
+	// in case we add new entires to null JSON objects. We don't want to do
```
Comment (rossf7) [inline pkg/app/mutate_app.go:104]:
```suggestion
	// in case we add new entries to null JSON objects. We don't want to do
```

**Review by rossf7 — APPROVED** [review]

File: pkg/app/mutate_app.go (hunk 1 above)
Comment (kopiczko) [reply pkg/app/mutate_app.go:83]: It is but it won't be dispatched as [we don't subscribe to it](https://github.com/giantswarm/app-admission-controller/blob/c614852d5787a6a7cb5141dd56ab190eb7181224/helm/app-admission-controller/templates/webhook.yaml#L26-L28). I may as well remove it.

**Pull request merged**